File: app/src/main/java/org/michaelbel/tjgram/ui/main/adapter/LikesDiffUtils.java
package org.michaelbel.tjgram.ui.main.adapter;

import android.os.Bundle;

import org.michaelbel.tjgram.data.entity.Entry;

import java.util.List;
import java.util.Objects;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;

import static org.michaelbel.tjgram.ui.main.adapter.EntriesAdapter.PAYLOAD_LIKE;

@SuppressWarnings("unused")
public class LikesDiffUtils extends DiffUtil.Callback {

    private List<Entry> oldList;
    private List<Entry> newList;

    LikesDiffUtils(List<Entry> oldList, List<Entry> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList != null ? oldList.size() : 0;
    }

    @Override
    public int getNewListSize() {
        return newList != null ? newList.size() : 0;
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        Entry oldEntry = oldList.get(oldItemPosition);
        Entry newEntry = newList.get(newItemPosition);
        return oldEntry.id == newEntry.id;
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        Entry oldEntry = oldList.get(oldItemPosition);
        Entry newEntry = newList.get(newItemPosition);
        return Objects.requireNonNull(oldEntry.likes).summ == Objects.requireNonNull(newEntry.likes).summ;
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        Entry newEntry = newList.get(newItemPosition);
        Entry oldEntry = oldList.get(oldItemPosition);

        boolean isLikesSame = Objects.requireNonNull(oldEntry.likes).summ == Objects.requireNonNull(newEntry.likes).summ;

        Bundle bundle = new Bundle();

        if (!isLikesSame) {
            bundle.putSerializable(PAYLOAD_LIKE, newEntry.likes);
        }

        return bundle.size() == 0 ? null : bundle;
    }
}

File: app/src/main/java/org/michaelbel/tjgram/ui/main/adapter/LikesDiffUtils.kt
package org.michaelbel.tjgram.ui.main.adapter

import android.os.Bundle
import androidx.recyclerview.widget.DiffUtil
import org.michaelbel.tjgram.data.entity.Entry
import org.michaelbel.tjgram.ui.main.adapter.EntriesAdapter.Companion.PAYLOAD_LIKE

class LikesDiffUtils internal constructor(private val oldList: List<Entry>, private val newList: List<Entry>) : DiffUtil.Callback() {

    override fun getOldListSize(): Int {
        return oldList.size
    }

    override fun getNewListSize(): Int {
        return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldEntry = oldList[oldItemPosition]
        val newEntry = newList[newItemPosition]
        return oldEntry.id == newEntry.id
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldEntry = oldList[oldItemPosition]
        val newEntry = newList[newItemPosition]
        return oldEntry.likes?.summ == newEntry.likes?.summ
    }

    override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
        val oldEntry = newList[newItemPosition]
        val newEntry = oldList[oldItemPosition]

        val isLikesSame = oldEntry.likes?.summ == newEntry.likes?.summ

        val bundle = Bundle()

        if (!isLikesSame) {
            bundle.putSerializable(PAYLOAD_LIKE, newEntry.likes)
        }

        return if (bundle.size() == 0) null else bundle
    }
}